### ISSUE TITLE
Replace Hector by Datastax Java Driver for Apache Cassandra

### ISSUE THREAD

**Issue by koppor** [kairosdb/kairosdb#228]
The last release of hector was on 2014-06-16 and the homepage states

> THIS PROJECT IS NO LONGER ACTIVE
> 
> Please use the official java-driver at https://github.com/datastax/java-driver/ for all Java-based Apache Cassandra projects. 
> [...]
> The currently active branch is 1.0. The master tracks Apache Cassandra active development which is 1.1.x presently.

The current release of Cassandra is 2.1.10. I don't know if that is too far away or if nothing important (from the point of the Java client side) has changed from 1.1.x to 2.1.10.

**Comment by lcoulet** [comment]
+1 for the move  and +1 to use the asynch features of datastax driver to increase retrieval performances.

But all of that as KairosDB doesn't migrate to awkward CQL data types and keeps its schema intact (or almost).

**Comment by gmucha** [comment]
+1 - I was able to do a quick POC for that - replacing just querying the data points. Good news is that this required no changes to the schema; even better news is the performance improvements I saw with that - I saw some queries' times cut in half using async statements, memory management also seems more stable.

**Comment by brianhks** [comment]
gmucha, please post your poc someplace.  We will have to make this part of the next release 1.0.2

**Comment by gmucha** [comment]
@brianhks  here it is. Very basic and probably doesn't support some features (like desc sorting), but seemed to work on my (basic) queries with same results as 1.0.0

https://gist.github.com/gmucha/e1d917233f5633679640

**Comment by rroopreddy** [comment]
We are using Cassandra 2.2.x and seems like "cassandra-cli" tooland connectivity to thrift port (9160)  is not available? Just wanted to play with KairosDB but may not be possible with current thrift only model?

**Comment by petergardfjall** [comment]
@rroopreddy 
Have you tried running with `start_rpc` set to `true` in `cassandra.yaml`?
That should enable the Thrift RPC server, which is disabled by default in Cassandra 2.2+.

https://issues.apache.org/jira/browse/CASSANDRA-9319

**Comment by noorul** [comment]
@gmucha Did you do any bench marking? Like, what is the speed of filtering data_points key from row_key_index for a given date range?

**Comment by kvlou** [comment]
Is this change planned to happen?

**Comment by brianhks** [comment]
Yes absolutely.  I'll need to put some other items on hold and get this done as C* is rolling out new versions that don't have thrift enabled.

**Comment by paladini** [comment]
The latest version of Cassandra don't have Thrift enabled by default :/ 

+1 to this feature. Do you need some help with it, @brianhks ? (also, can you add me as a collaborator of the KairosDB organization? I would like to create two repositories [kairosdb-examples-python and kairosdb-examples-R]).

**Comment by brianhks** [comment]
I'm working on it now.  I would like some help testing when it is done.  @paladini I sent you an invite via github, check your inbox.

**Comment by paladini** [comment]
Okay, I can help testing this new feature.

**Comment by paladini** [comment]
[off-topic, will delete it later] @brianhks I've accepted the request, but still can't create new repositories on the KairosDB organization. Can you give me permissions for that?

**Comment by Weijianlin** [comment]
It's not easy to replace the driver.  You have to change the schema because the cql changes the storage.      
You can not create a table schema with 2b columns. You have to use the map collection, and it limits the values to 64K.

**Comment by uhoh-itsmaciek** [comment]
@brianhks also happy to help test when this is ready

**Comment by lcoulet** [comment]
Actually it is relatively easy. The current schema IS cql compliant. Just
it doesn't use cql data types - all blobs.

Why do you think it's not?

**Comment by paladini** [comment]
As I've said before in the KairosDB forum, I would like to create a truly integration between KairosDB and Spark (because I currently really need something like that in my laboratory). My first intention was to simply query the KairosDB HTTP API, but this will take a lot of precious time and Spark already has some integrations with Cassandra. If you move from Hector to Datastax Java Driver (and CQL), I can an idea on how KairosDB stores it's data on Cassandra. 

@brianhks , as soon as you get it done I'll start working in a project like the above one.

**Comment by gmucha** [comment]
WRT/benchmarking - I didn't update the ingestion, so it was still storing data using Thrift, but querying via CQL; using async queries with a 5-node cluster I was seeing roughly 2x faster query responses. Looking at read speed Thrift was able to do around 300k datapoints read/second, Datastax driver was roughly 600k points/second. I didn't do much further optimization, but the disk-based query cache would probably be the next point for optimization). Also - for a longer time horizont it waited for the stragglers from the currently queried timestamp generation before advancing forward in time.  

I was also wondering if the async nature of CQL driver wouldn't lend itself to more effective result processing - since you can keep multiple queries open/running and you already have the time series metadata (from the index table) it would be theoretically possible to read data and perform the aggregations on the fly - without ever going via disk. 

@paladini If you connect to the kairosdb namespace in Cassandra you can see how the tables are structured. Generally you'd have to reimplement the QueryRunner class (like the CQLQueryRunner in my gist) + the logic to identify rows to query via  the index table. The gist of the query engine is actually getting rows from the index table based on the metadata( metric name, tags etc and row start time) and doing a multi-get (or multiple CQL SELECTs) with range scans on the columns identified.

**Comment by Weijianlin** [comment]
for branch 'feature/cql',  it misuses the batch statement!  Improved performance is not a
reason to use batches!

I fork kairosdb, and already use Datastax driver instead of Hector.  But I use counter column to redesign the schema for my project.

**Comment by Weijianlin** [comment]
@lcoulet   sorry

I check the wrong branch 'cql'.  The schema below is wrong. It limits 2 billion data points.

but the 'feature/cql' is right.

`session.execute("CREATE TABLE timeseries.data_points (" +
					"metric_name text, " +
					"value int, " +
					"time timestamp, " +
					"tags text, " +
					"PRIMARY KEY ((metric_name, tags), time)) WITH COMPACT STORAGE;");`

**Comment by Jan-M** [comment]
Is there any progress on this issue? getting rid of hector would really be great.

You need any help? Testing with bigger deployment is def. possible on our side.

We run into some issues here and there that seem to stem more from hector than KairosDB / Cassandra itself. Something we don't see with software running the Datastax driver.

**Comment by paladini** [comment]
I have millions of point values that I can use for testing purposes too. If you need any help on testing, please let us know. 

One more question: using DataStax driver will allow us to create tools to KairosDB easily (tools like a KairosDB-Spark connector and KairosDB shell)?

**Comment by Jan-M** [comment]
Out of curiosity I did deploy feature/cql into our staging environment, basically this did work in terms of querying data.

Rolling it out into production seemed a bit more flaky with the traffic it receives (I did remove the possibly broken batching in the write path). But for smaller use cases this maybe worth trying more.

Ingest time:
![image](https://cloud.githubusercontent.com/assets/1443395/14431634/c9c17944-0006-11e6-833d-7ce052efd0fb.png)

Ingest count:
![image](https://cloud.githubusercontent.com/assets/1443395/14431663/ee9bccd8-0006-11e6-9355-81533b05a485.png)

(where red and blue where the cql branch)

I will take a deeper look at this in due time.

**Comment by Jan-M** [comment]
Looking more into this, I have now understood that there is still code using Hector, which I am rewriting now.

Also the Cassandra Session was using a hard coded consistency level of quorum, which I have changed now. We are actually running one for read and write, so I will retry this later.

**Comment by paladini** [comment]
Wonderful work, @Jan-M ! Please, let us know if you achieve the driver replacement! Of course, try to merge your code into KairosDB project. 

We're really expecting that in my lab. Do you need some help? :)

**Comment by Jan-M** [comment]
If I am not mistaken, the feature/cql branch actually only converted the write path so far. I mixed this up by reading this thread and the gist. Anyways I will see to merging this with the stuff from the gist (sadly I spent some time now writing exactly that stuff, but good for understanding anyways).

**Comment by Jan-M** [comment]
The below branch is now roled out in our prod environment for testing.

 * write now uses CQL to my understanding
 * reading row key index and string index is now also using CQL
 * changed quorum to 1

https://github.com/Jan-M/kairosdb/tree/feature/cql

However, ingest_time metric now is pretty bad, this most likely due to not having write buffer for now.

![image](https://cloud.githubusercontent.com/assets/1443395/14465770/744974d4-00d4-11e6-9f83-68e762e9ecd5.png)

Query time I cannot spot any difference (no change there). Key Query time looks lower (in chart the lowest ones at the bottom dropping to 0 imho).

![image](https://cloud.githubusercontent.com/assets/1443395/14465881/e1291320-00d4-11e6-9bfb-3ee97fffb349.png)

Maybe someone wants to look at the CQLRunner (I put the stuff from the gist in there ) integration, sadly I am out of time for today and the next two days.

**Comment by Jan-M** [comment]
FYI: There branch is now up to date and also uses CQL to read data points. But it is not worth using yet, performance is much worse as I tried to model it really close to the original QueryRunner and also had to remove the fetch for multiple row keys.

The code from the gist with the async stuff is also in there, named in AsyncCQLRunner for now, but one should look think about the suggested Singleton vs Synchronized, this would not work.

I am also wondering about the need for the cached search result to rely on files, in our environment we have about 10GB of cache files and most caches are only valid for one hour, and getting random writes to the "searchresult" over having to write tags sequentially would make stuff easier (RAM is cheap now adays).

**Comment by Jan-M** [comment]
I updated the branch to use ResultSetFuture for every RowKey to query and this speed up the result significantly on our staging environment. But compared to the production env. it does not look faster yet.

**Comment by gmucha** [comment]
Well - the main gain is to try and process the futures as they become
available - right now the code in the branch is launching all the queries -
but also retrieving them in the same order. That's why the gist I sent was
using ListenableFutures - so that whatever query got processed first,gets
written into the cache file. This way you can pipeline the writes (via
writeColumnsCQL).

Since the cache is indexed, within a single timestamp window (ie. 3 weeks)
the order of tags shouldn't matter.

Now for the cache (although that's a completely different matter). I was
wondering if - having read the row key indexes - this information couldn't
be use to query the data rows in parallel - ie. keep multiple Cassandra
queries running and traverse the cursors in parallel. It may not be
possible for massive queries (ie. thousands of rows) and it won't work for
certain aggregates


2016-04-15 15:40 GMT+02:00 Jan Mussler <notifications@github.com>:

**Comment by Jan-M** [comment]
I need to put some more thoughts into this, it looks like there is some stuff not handled right, the async version does not work 100% (results differ), I was to eager here.

Yes, I agree with the listenable but trying to do one step at a time, the benefit probably also largely depends on how much time the postprocessing actually takes.

Will continue next week though, for me also if performance matches hector somewhat I would still prefer having kairosdb rely on the CQL driver.

**Comment by Jan-M** [comment]
I did fix the async implementation, I missed that BoundStatement can only be reused with the sync api. Now results look fine again.

As far as the listenableFuture impl. goes this is still open. So is removing delete hector code and configuration code.

My personal branch is now also deployed, but so far no big change visible.

**Comment by paralin** [comment]
Any updates on this? @Jan-M

**Comment by brianhks** [comment]
Hey guys, sorry this thread dropped off my radar.  If you have been following the discussion group I have been posting results of the CQL conversion and it is under active development.

**Comment by pauldub** [comment]
Hi @brianhks can I ask you what is the progress on the `feature/cql` branch? A quick glance tells me it looks quite complete already.

I'd be willing to contribute a few minor patches (handle ports in cql_host_list, pass credentials to the cluster builder, etc.) if you are interested.

**Comment by francisdb** [comment]
Is there a milestone for v1.2.0 final?

**Comment by brianhks** [comment]
1.2 beta is out and all CQL functionality has been implemented.  Should have a final release by the end of the year.

**Comment by francisdb** [comment]
@brianhks mind to elaborate a bit on the mentioned 8x speed improvement?
